pkg/controllers/applicationfailover/crb_application_failover_controller.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/karmada-io/karmada/pkg/resourceinterpreter"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/keys"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+// CRBApplicationFailoverControllerName is the controller name that will be used when reporting events.
+const CRBApplicationFailoverControllerName = "cluster-resource-binding-application-failover-controller"
+
+// CRBApplicationFailoverController is to sync ClusterResourceBinding's application failover behavior.
+type CRBApplicationFailoverController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+
+	// workloadUnhealthyMap records which clusters the specific resource is in an unhealthy state
+	workloadUnhealthyMap *workloadUnhealthyMap
+	ResourceInterpreter  resourceinterpreter.ResourceInterpreter
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *CRBApplicationFailoverController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling ClusterResourceBinding %s.", req.Name)
+
+	binding := &workv1alpha2.ClusterResourceBinding{}
+	if err := c.Client.Get(ctx, req.NamespacedName, binding); err != nil {
+		if apierrors.IsNotFound(err) {
+			resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+			if err != nil {
+				return controllerruntime.Result{Requeue: true}, err
+			}
+			c.workloadUnhealthyMap.delete(resource)
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if !c.clusterResourceBindingFilter(binding) {
+		resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+		if err != nil {
+			return controllerruntime.Result{Requeue: true}, err
+		}
+		c.workloadUnhealthyMap.delete(resource)
+		return controllerruntime.Result{}, nil
+	}
+
+	retryDuration, err := c.syncBinding(binding)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if retryDuration > 0 {
+		klog.V(4).Infof("Retry to check health status of the workload after %v minutes.", retryDuration.Minutes())
+		return controllerruntime.Result{RequeueAfter: retryDuration}, nil
+	}
+	return controllerruntime.Result{}, nil
+}
+
+func (c *CRBApplicationFailoverController) detectFailure(clusters []string, tolerationSeconds *int32, resource keys.ClusterWideKey) (int32, []string) {
+	var needEvictClusters []string
+	var duration int32
+
+	for _, cluster := range clusters {
+		if !c.workloadUnhealthyMap.hasWorkloadBeenUnhealthy(resource, cluster) {
+			c.workloadUnhealthyMap.setTimeStamp(resource, cluster)
+			duration = *tolerationSeconds
```

Comment:
I prefer to add continue, and delete `else`, which will make `Cyclomatic complexity` smaller.

Reply:
+1

Reply:
+1